src/nvim/eval/decode.c

```diff
@@ -0,0 +1,945 @@
+          assert(last_container.stack_index < kv_size(stack) - 1);
+          EMSG2(_("E474: Expected value: %s"), p);
+          goto json_decode_string_fail;
+        }
+        if (kv_size(stack) == 1) {
+          p++;
+          (void) kv_pop(container_stack);
+          goto json_decode_string_after_cycle;
+        } else {
+          if (json_decoder_pop(kv_pop(stack), &stack, &container_stack, &p,
+                               &next_map_special, &didcomma, &didcolon)
+              == FAIL) {
+            goto json_decode_string_fail;
+          }
+          assert(!next_map_special);
+          break;
+        }
+      }
+      case ',': {
+        if (kv_size(container_stack) == 0) {
+          EMSG2(_("E474: Comma not inside container: %s"), p);
+          goto json_decode_string_fail;
+        }
+        ContainerStackItem last_container = kv_last(container_stack);
+        if (didcomma) {
+          EMSG2(_("E474: Duplicate comma: %s"), p);
+          goto json_decode_string_fail;
+        } else if (didcolon) {
+          EMSG2(_("E474: Comma after colon: %s"), p);
+          goto json_decode_string_fail;
+        } else if (last_container.container.v_type == VAR_DICT
+                   && last_container.stack_index != kv_size(stack) - 1) {
+          EMSG2(_("E474: Using comma in place of colon: %s"), p);
+          goto json_decode_string_fail;
+        } else if (last_container.special_val == NULL
+                   ? (last_container.container.v_type == VAR_DICT
+                      ? (DICT_LEN(last_container.container.vval.v_dict) == 0)
+                      : (last_container.container.vval.v_list->lv_len == 0))
+                   : (last_container.special_val->lv_len == 0)) {
+          EMSG2(_("E474: Leading comma: %s"), p);
+          goto json_decode_string_fail;
+        }
+        didcomma = true;
+        continue;
+      }
+      case ':': {
+        if (kv_size(container_stack) == 0) {
+          EMSG2(_("E474: Colon not inside container: %s"), p);
+          goto json_decode_string_fail;
+        }
+        ContainerStackItem last_container = kv_last(container_stack);
+        if (last_container.container.v_type != VAR_DICT) {
+          EMSG2(_("E474: Using colon not in dictionary: %s"), p);
+          goto json_decode_string_fail;
+        } else if (last_container.stack_index != kv_size(stack) - 2) {
+          EMSG2(_("E474: Unexpected colon: %s"), p);
+          goto json_decode_string_fail;
+        } else if (didcomma) {
+          EMSG2(_("E474: Colon after comma: %s"), p);
+          goto json_decode_string_fail;
+        } else if (didcolon) {
+          EMSG2(_("E474: Duplicate colon: %s"), p);
+          goto json_decode_string_fail;
+        }
+        didcolon = true;
+        continue;
+      }
+      case ' ':
+      case TAB:
+      case NL: {
+        continue;
+      }
+      case 'n': {
+        if (strncmp(p + 1, "ull", 3) != 0) {
```

Comment:
Is `p` NUL terminated? Otherwise the following byte positions could already be outside of the buffer.